File: samples/system/typescript/07_sessionStickiness.ts
/**
 * 07 - Session Stickiness (Conversational Affinity)
 *
 * Configures session-sticky routing so that all requests within the same
 * conversation are routed to the same model. This sample demonstrates:
 *
 *   - Session-stickiness rotation policy
 *   - Sending multiple messages with the same session_id
 *   - Verifying that all messages route to the same model
 *   - Fallback behavior when the session's model becomes unavailable
 *   - Multiple independent sessions routed independently
 *
 * Session stickiness is important for multi-turn conversations where
 * switching models mid-conversation can cause inconsistency in tone,
 * context interpretation, or response style.
 *
 * Prerequisites:
 *   - Set OPENAI_API_KEY and ANTHROPIC_API_KEY environment variables
 */

import { ModelMesh, MeshConfig } from "modelmesh-lite";

async function main(): Promise<void> {
  // -----------------------------------------------------------------------
  // 1. Configure with session-stickiness policy
  // -----------------------------------------------------------------------
  const config: MeshConfig = {
    raw: {
      providers: {
        "openai.llm.v1": {
          enabled: true,
          api_key: "${secrets:OPENAI_API_KEY}",
        },
        "anthropic.llm.v1": {
          enabled: true,
          api_key: "${secrets:ANTHROPIC_API_KEY}",
        },
      },

      pools: {
        "text-generation": {
          // Session-stickiness routes all requests with the same session_id
          // to the same model. The first request in a session picks a model
          // using the fallback strategy; subsequent requests reuse it.
          strategy: "modelmesh.session-stickiness.v1",
          fallback_strategy: "modelmesh.round-robin.v1",

          deactivation: {
            retry_limit: 3,
            error_codes: [429, 500, 503],
          },
          recovery: {
            cooldown: "30s",
          },
          retry: {
            max_attempts: 2,
            backoff: "exponential_jitter",
            initial_delay: "500ms",
            retryable_codes: [429, 500, 502, 503],
            scope: "same_model",
          },
        },
      },

      observability: {
        routing: {
          connector: "modelmesh.console.v1",
        },
      },
    },
  };

  // -----------------------------------------------------------------------
  // 2. Initialize
  // -----------------------------------------------------------------------
  const mesh = new ModelMesh();
  await mesh.initialize(config);
  console.log("ModelMesh initialized with session-stickiness routing.\n");

  const client = mesh.getClient();
  const router = mesh.getRouter();

  // -----------------------------------------------------------------------
  // 3. Session A: multi-turn conversation (same session_id)
  // -----------------------------------------------------------------------
  const sessionA = "session-user-alice-001";
  console.log(`--- Session A (${sessionA}) ---\n`);

  // The RoutingOptions.sessionId is passed through the router. When using
  // the OpenAI-compatible client, the session_id can be set in metadata
  // or through the routing options on the underlying router.

  // Turn 1
  const responseA1 = await client.chat.completions.create({
    model: "text-generation",
    messages: [
      { role: "user", content: "Hi! My name is Alice. Remember it." },
    ],
    max_tokens: 100,
    // Session ID is passed as extra metadata for the routing layer
    session_id: sessionA,
  } as any);

  console.log(`Turn 1 -> Model: ${responseA1.model}`);
  console.log(`         Reply : ${(responseA1.choices[0] as any).message.content}\n`);

  // Turn 2 (should route to the SAME model)
  const responseA2 = await client.chat.completions.create({
    model: "text-generation",
    messages: [
      { role: "user", content: "Hi! My name is Alice. Remember it." },
      { role: "assistant", content: (responseA1.choices[0] as any).message.content },
      { role: "user", content: "What is my name?" },
    ],
    max_tokens: 100,
    session_id: sessionA,
  } as any);

  console.log(`Turn 2 -> Model: ${responseA2.model}`);
  console.log(`         Reply : ${(responseA2.choices[0] as any).message.content}\n`);

  // Turn 3 (should still route to the SAME model)
  const responseA3 = await client.chat.completions.create({
    model: "text-generation",
    messages: [
      { role: "user", content: "Hi! My name is Alice. Remember it." },
      { role: "assistant", content: (responseA1.choices[0] as any).message.content },
      { role: "user", content: "What is my name?" },
      { role: "assistant", content: (responseA2.choices[0] as any).message.content },
      { role: "user", content: "Tell me a joke about my name." },
    ],
    max_tokens: 150,
    session_id: sessionA,
  } as any);

  console.log(`Turn 3 -> Model: ${responseA3.model}`);
  console.log(`         Reply : ${(responseA3.choices[0] as any).message.content}\n`);

  // Verify session consistency
  const sameModel = responseA1.model === responseA2.model && responseA2.model === responseA3.model;
  console.log(`Session A consistency: ${sameModel ? "All 3 turns routed to the same model" : "Model changed mid-session (failover occurred)"}`);

  // -----------------------------------------------------------------------
  // 4. Session B: independent session (may route to a different model)
  // -----------------------------------------------------------------------
  const sessionB = "session-user-bob-002";
  console.log(`\n--- Session B (${sessionB}) ---\n`);

  const responseB1 = await client.chat.completions.create({
    model: "text-generation",
    messages: [
      { role: "user", content: "Hello, I am Bob. What provider are you?" },
    ],
    max_tokens: 100,
    session_id: sessionB,
  } as any);

  console.log(`Turn 1 -> Model: ${responseB1.model}`);
  console.log(`         Reply : ${(responseB1.choices[0] as any).message.content}\n`);

  console.log("Note: Session A and Session B may use different models,");
  console.log("demonstrating that sessions are routed independently.");
  console.log(`Session A model: ${responseA1.model}`);
  console.log(`Session B model: ${responseB1.model}`);

  // -----------------------------------------------------------------------
  // 5. Shut down
  // -----------------------------------------------------------------------
  await mesh.shutdown();
  console.log("\nModelMesh shut down.");
}

main().catch(console.error);

File: samples/system/python/07_session_stickiness.py
"""
07 - Session Stickiness (Conversational Affinity)
==================================================

Session-sticky routing ensures all messages within a conversation are sent to
the same model and provider.  This is important for multi-turn conversations
where consistent model behavior is desired.

This sample demonstrates:
  - Configuring session-stickiness routing
  - Multi-turn conversations using the same model
  - Verifying that all messages in a session route to the same model

Uses mock providers so it runs without API keys.
"""

import asyncio
import uuid
from modelmesh import ModelMesh, MeshConfig
from modelmesh.interfaces.provider import (
    ChatMessage, CompletionChoice, CompletionRequest, CompletionResponse,
    ErrorClassification, ModelInfo, ModelPricing, ProviderConnector,
    QuotaStatus, RateLimitStatus, TokenUsage,
)


class MockConversationProvider(ProviderConnector):
    """Mock provider that tracks conversation context."""

    def __init__(self, name: str):
        self._name = name

    async def complete(self, request: CompletionRequest) -> CompletionResponse:
        user_msg = ""
        for msg in request.messages:
            if isinstance(msg, dict) and msg.get("role") == "user":
                user_msg = msg.get("content", "")

        reply = f"[{self._name}] I understand. Regarding '{user_msg[:40]}...' -- here is my answer."
        return CompletionResponse(
            id=f"resp-{self._name}",
            model=self._name,
            choices=[CompletionChoice(
                index=0,
                message=ChatMessage(role="assistant", content=reply),
                finish_reason="stop",
            )],
            usage=TokenUsage(prompt_tokens=15, completion_tokens=20, total_tokens=35),
        )

    async def stream(self, request):
        yield await self.complete(request)

    def get_capabilities(self):
        return ["generation.text-generation.chat-completion"]

    def supports(self, cap):
        return cap in self.get_capabilities()

    def list_models(self):
        return [ModelInfo(id=self._name, name=self._name)]

    def get_model_info(self, mid):
        return ModelInfo(id=mid, name=mid)

    def check_quota(self):
        return QuotaStatus()

    def get_rate_limits(self):
        return RateLimitStatus()

    def get_pricing(self, mid):
        return ModelPricing()

    def report_usage(self, mid, usage):
        pass

    def classify_error(self, error):
        return ErrorClassification(retryable=False)


async def run_conversation(client, session_id: str, label: str) -> None:
    """Simulate a multi-turn conversation within a single session."""
    print(f"\n{'=' * 50}")
    print(f"Session: {label} (id: {session_id[:8]}...)")
    print(f"{'=' * 50}")

    messages = [
        {"role": "system", "content": "You are a helpful tutor. Be concise."},
    ]

    turns = [
        "What is recursion in programming?",
        "Can you give me a simple example in Python?",
        "What are the risks of using recursion without a base case?",
    ]

    for i, user_msg in enumerate(turns, start=1):
        messages.append({"role": "user", "content": user_msg})

        print(f"\n  Turn {i}: {user_msg}")

        response = client.chat.completions.create(
            model="text-generation",
            messages=messages,
            temperature=0.5,
            max_tokens=200,
        )

        assistant_msg = response.choices[0].message.content
        messages.append({"role": "assistant", "content": assistant_msg})

        print(f"  Model  : {response.model}")
        print(f"  Answer : {assistant_msg[:100]}...")


async def main() -> None:
    config = MeshConfig(raw={
        "providers": {
            "openai": {
                "connector": "openai",
                "enabled": True,
                "instance": MockConversationProvider("gpt-4o"),
            },
            "anthropic": {
                "connector": "anthropic",
                "enabled": True,
                "instance": MockConversationProvider("claude-sonnet-4"),
            },
        },
        "models": {
            "openai.gpt-4o": {
                "provider": "openai",
                "capabilities": ["generation.text-generation.chat-completion"],
            },
            "anthropic.claude-sonnet-4": {
                "provider": "anthropic",
                "capabilities": ["generation.text-generation.chat-completion"],
            },
        },
        "pools": {
            "text-generation": {
                "capability": "generation.text-generation.chat-completion",
                "strategy": "stick-until-failure",
            },
        },
    })
    mesh = ModelMesh()
    mesh.initialize(config)
    client = mesh.get_client()

    print("Session-stickiness routing demo")
    print("-" * 40)
    print("Two independent conversations run with different session IDs.")
    print("Each conversation is pinned to whichever model handles its")
    print("first request.\n")

    session_a = str(uuid.uuid4())
    session_b = str(uuid.uuid4())

    await run_conversation(client, session_a, "Conversation A")
    await run_conversation(client, session_b, "Conversation B")

    print(f"\n{'=' * 50}")
    print("Fallback scenario")
    print(f"{'=' * 50}")
    print()
    print("If the model pinned to a session becomes unavailable,")
    print("ModelMesh will retry and then rotate to another model,")
    print("re-pinning the session to the new model going forward.")

    mesh.shutdown()
    print("\nModelMesh shut down.")


if __name__ == "__main__":
    asyncio.run(main())
